### PR TITLE
Basic version of test requirement scripts

### PR DESCRIPTION
**Pull request by bendudson**
Set of small scripts used to test individual requirements. These can then be combined using python expressions in a requirements specification in the runtest script. Adding a new requirement check is done by putting an executable script in the "requirements" directory, which is then automatically run by the test suite.

Currently only added to test-multigrid_laplace, to skip when travis is running with mpich:
`#requires not (travis and mpich)`

Possible solution to issue #979, alternative to PR #943

### CODE REVIEW

File: bin/bout-config.in [hunk 1]
```diff
@@ -24,7 +24,6 @@ libs="@CONFIG_LDFLAGS@"
 
 version="@BOUT_VERSION@"
 git="@GIT_REVISION@"
-checks="@CHECK_LEVEL@"
```
Comment (bendudson) [inline bin/bout-config.in]: Is this still needed? I think it's printed further down the file

Comment (dschwoerer) [reply bin/bout-config.in]: No, its printed a bit further up - that's why I removed this duplicate.

**Comment by dschwoerer** [comment]
This looks probably nicer than what I started with :+1: 
I have extended that a bit - I hope you are happy with the changes.

Are you happy if I include the utf-8 symbols again here?

**Comment by bendudson** [comment]
Yes, thanks! It's a bit more complicated now, but quite nicely wrapped up. A nice bonus to be able to replace the `!` notation. 

which utf-8 symbols? Yes, happy to include utf-8 everywhere...

**Comment by bendudson** [comment]
Did you mean to add some requirement scripts? One of the commit messages contains
"Get infos about whether we compiled with netcdf/hdf5"

**Comment by dschwoerer** [comment]
I meant commit bde32b58fc94a4470401c2cf17ec15b68e34c682 (now 0c8f5909d5558a547c294c7a7791ce877f6df00b) - which prints ✓ and ❌ instaed of . and F

The checks for netcdf etc are not the way you intended them - they are added to the dictionary by reading the bout-config script.
I thought that it would be easier then to write scripts for each case - in the end we don't really want to know whether we have netcdf (which is part of the tests), but rather whether the build should have netcdf.
This is done [here](https://github.com/boutproject/BOUT-dev/pull/980/commits/6206d1dd6b002ac712e866607bcc55bfec205164#diff-d0126c77d6847a14ee57f11a2d46a492R30)

**Comment by ZedThree** [comment]
Just had a quick glance over this, it looks good.

A couple of things:
- Would having `conflicts` as a synonym for `requires not` be useful, or just make it more complicated?
- The `Requirements` class should be moved out of `__init__.py` -- that's just really for module initialisation -- into a separate file

**Comment by dschwoerer** [comment]
Adding `conflicts` to the parser would be easy - I am however not sure it makes it easier to read:
`#conflicts: mpich and travis` sounds like it would conflict with ether mpich or travis (at least from normal day language usage).

Besides that, I moved Requirements to it's own class - hope that is done in the correct way.

**Comment by dschwoerer** [comment]
Is this ready for merging?
If so I would merge this into #641 that the cython branch gets ready for merging ...

**Comment by bendudson** [comment]
Yes,  ready to merge I think.  Unfortunately I can't resolve conflicts on a phone

**Pull request merged**